TYP return type of DataFrame.clip not correct

https://github.com/pandas-dev/pandas/blob/094f630cd84365b1902b4ee8d3de6a72d1bcbca3/pandas/core/generic.py#L7449-L7458

Return type should be `Optional[FrameOrSeriesUnion]`

It could also do with being overloaded according to the value of `inplace` - see `set_axis` and `DataFrame.reset_index` for examples

@MarcoGorelli Did you mean that the return type is supposed to be  `FrameOrSeriesUnion | None`  or `FrameOrSeries | None` ? If I use  FrameOrSeriesUnion, I get `pandas/core/generic.py:7538: error: Incompatible return value type (got "FrameOrSeries", expected "Optional[Union[DataFrame, Series]]")  [return-value]` On line 7526, `result = self`. So `FrameOrSeries` seems applicable here?

--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -7383,6 +7383,58 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
     def clip(
         self: FrameOrSeries,
         lower,
+        *,
+        axis,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        lower,
+        *,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        *,
+        upper,
+        axis,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        *,
+        upper,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        *,
+        axis,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        lower,
         upper,
         axis: Axis | None,
         inplace: Literal[True],
@@ -7396,6 +7448,15 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
         self: FrameOrSeries,
         lower,
         upper,
+        *,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
         *,
         inplace: Literal[True],
         **kwargs,

--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -7384,7 +7384,7 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
         self: FrameOrSeries,
         lower,
         *,
-        axis,
+        axis: Axis | None,
         inplace: Literal[True],
         **kwargs,
     ) -> None:
@@ -7405,7 +7405,7 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
         self: FrameOrSeries,
         *,
         upper,
-        axis,
+        axis: Axis | None,
         inplace: Literal[True],
         **kwargs,
     ) -> None:
@@ -7425,7 +7425,7 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
     def clip(
         self: FrameOrSeries,
         *,
-        axis,
+        axis: Axis | None,
         inplace: Literal[True],
         **kwargs,
     ) -> None:

--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -33,6 +33,7 @@ from pandas._libs.tslibs import (
     to_offset,
 )
 from pandas._typing import (
+    ArrayLike,
     Axis,
     CompressionOptions,
     Dtype,
@@ -7367,16 +7368,52 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
                 threshold = align_method_FRAME(self, threshold, axis, flex=None)[1]
         return self.where(subset, threshold, axis=axis, inplace=inplace)
 
-    @final
+    @overload
     def clip(
         self: FrameOrSeries,
-        lower=None,
-        upper=None,
-        axis=None,
-        inplace: bool_t = False,
+        inplace: Literal[False] = ...,
+        lower: int | ArrayLike | None = None,
+        upper: int | ArrayLike | None = None,
+        axis: Axis | None = None,
         *args,
         **kwargs,
     ) -> FrameOrSeries:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        inplace: Literal[True],
+        lower: int | ArrayLike | None = None,
+        upper: int | ArrayLike | None = None,
+        axis: Axis | None = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        inplace: bool_t = ...,
+        lower=None,
+        upper=None,
+        axis: Axis | None = None,
+        *args,
+        **kwargs,
+    ) -> FrameOrSeries | None:
+        ...
+
+    @final
+    def clip(
+        self: FrameOrSeries,
+        inplace: bool_t = False,
+        lower=None,
+        upper=None,
+        axis: Axis | None = None,
+        *args,
+        **kwargs,
+    ) -> FrameOrSeries | None:
         """
         Trim values at input threshold(s).
 

--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -33,7 +33,6 @@ from pandas._libs.tslibs import (
     to_offset,
 )
 from pandas._typing import (
-    ArrayLike,
     Axis,
     CompressionOptions,
     Dtype,
@@ -7371,10 +7370,10 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
     @overload
     def clip(
         self: FrameOrSeries,
+        lower=...,
+        upper=...,
+        axis: Axis | None = ...,
         inplace: Literal[False] = ...,
-        lower: int | ArrayLike | None = None,
-        upper: int | ArrayLike | None = None,
-        axis: Axis | None = None,
         *args,
         **kwargs,
     ) -> FrameOrSeries:
@@ -7383,10 +7382,10 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
     @overload
     def clip(
         self: FrameOrSeries,
+        lower,
+        upper,
+        axis: Axis | None,
         inplace: Literal[True],
-        lower: int | ArrayLike | None = None,
-        upper: int | ArrayLike | None = None,
-        axis: Axis | None = None,
         *args,
         **kwargs,
     ) -> None:
@@ -7395,10 +7394,21 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
     @overload
     def clip(
         self: FrameOrSeries,
+        lower,
+        upper,
+        *,
+        inplace: Literal[True],
+        **kwargs,
+    ) -> None:
+        ...
+
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        lower=...,
+        upper=...,
+        axis: Axis | None = ...,
         inplace: bool_t = ...,
-        lower=None,
-        upper=None,
-        axis: Axis | None = None,
         *args,
         **kwargs,
     ) -> FrameOrSeries | None:
@@ -7407,10 +7417,10 @@ class NDFrame(PandasObject, SelectionMixin, indexing.IndexingMixin):
     @final
     def clip(
         self: FrameOrSeries,
-        inplace: bool_t = False,
         lower=None,
         upper=None,
         axis: Axis | None = None,
+        inplace: bool_t = False,
         *args,
         **kwargs,
     ) -> FrameOrSeries | None:

